pkg/clusteragent/custommetrics/provider.go

```diff
@@ -34,24 +36,95 @@ type datadogProvider struct {
 	client dynamic.Interface
 	mapper apimeta.RESTMapper
 
-	values          map[provider.CustomMetricInfo]int64
 	externalMetrics []externalMetric
 	resVersion      string
 	store           Store
+	serving         bool
 	timestamp       int64
 	maxAge          int64
 }
 
 // NewDatadogProvider creates a Custom Metrics and External Metrics Provider.
-func NewDatadogProvider(client dynamic.Interface, mapper apimeta.RESTMapper, store Store) provider.MetricsProvider {
+func NewDatadogProvider(ctx context.Context, client dynamic.Interface, mapper apimeta.RESTMapper, store Store) provider.MetricsProvider {
 	maxAge := config.Datadog.GetInt64("external_metrics_provider.local_copy_refresh_rate")
-	return &datadogProvider{
+	d := &datadogProvider{
 		client: client,
 		mapper: mapper,
-		values: make(map[provider.CustomMetricInfo]int64),
 		store:  store,
 		maxAge: maxAge,
 	}
+	go d.externalMetricsSetter(ctx)
+	return d
+}
+
+type timer struct {
+	period time.Duration
+	timer  time.Timer
+}
+
+func (t *timer) resetTimer() {
+	t.timer.Stop()
+	t.timer = *time.NewTimer(t.period)
+}
+
+func createTimer(period time.Duration) *timer {
+	return &timer{period, *time.NewTimer(period)}
+}
+
+func (p *datadogProvider) externalMetricsSetter(ctx context.Context) {
+	log.Infof("Starting async loop to collect External Metrics")
+	tick := time.NewTicker(time.Duration(p.maxAge) * time.Second)
+	defer tick.Stop()
+	out := createTimer(2 * time.Duration(p.maxAge) * time.Second)
```

Comment:
I think you can use https://golang.org/pkg/context/#WithTimeout here

Reply:
The timer allows me to restart at the end of the tick. With a context timeout I would not be able to do that.
I am adding a WithCancel though, so I can tear everything down in case of permafail accessing the CM (3 retries).